### PR TITLE
Basic and token authentication support

### PR DESCRIPTION
**Pull request by josephglanville**
Implements support for "basic" and "token" authentication methods.

Partial implementation of what is called for in #49, to go the next step will need to extract the role from the authentication provider and then use that to authorise actions on the connection.

### CODE REVIEW

**Comment by josephglanville** [comment]
Similar to #54 I could do with some guidance here on if it's best to have authentication for mqtt enabled independently of the pulsar `authenticationEnabled` setting or not.

**Review by codelipenghui — COMMENTED** [review]
The change looks good, could you please also help update the README for how to enable authentication for mop?

**Comment by codelipenghui** [comment]
@josephglanville I noticed there are some checkstyle violations, please check, thanks.

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/ProtocolMethodProcessorImpl.java [hunk 1]
```diff
@@ -120,6 +129,34 @@ public void processConnect(Channel channel, MqttConnectMessage msg) {
...
+        if (!configuration.isMqttEnableAuthentication()) {
+            log.info("Authentication is disabled, allowing client. CId={}, username={}", clientId, payload.userName());
+        } else {
+            boolean authenticated = false;
+            for (String authMethod : configuration.getMqttAuthenticationMethods()) {
+                AuthenticationProvider authProvider = authService.getAuthenticationProvider(authMethod);
+                try {
+                    AuthenticationDataSource authData = getAuthData(authMethod, payload);
+                    if (authProvider != null && authData != null) {
+                        authService.getAuthenticationProvider(authMethod)
```
Comment (zymap) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/ProtocolMethodProcessorImpl.java:142]: duplicate with line 138?

**Review by zymap — COMMENTED** [review]
Nice contribute! Overall looks good

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/ProtocolMethodProcessorImpl.java (hunk 1 above)
Comment (josephglanville) [reply mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/ProtocolMethodProcessorImpl.java:142]: Yep, vestigial from different approach, will remove.

**Review by codelipenghui — APPROVED** [review]

**Comment by josephglanville** [comment]
@codelipenghui I have refactored this to not do the configuration in the ChannelInitialiser and rather ensure appropriate authentication providers and methods are configured at start time. I also added support to proxy connections to use authetnication.

Main things to do now are:
Tests for proxy auth
Failure tests (invalid creds etc)
Tests for token authentication.

**Comment by codelipenghui** [comment]
@josephglanville Sorry for the late response, I have merged the PR.

**Pull request merged**